geode-redis/src/distributedTest/java/org/apache/geode/redis/internal/executor/hash/HExistsDUnitTest.java

```diff
@@ -0,0 +1,151 @@
+import org.junit.Test;
+import redis.clients.jedis.Jedis;
+
+import org.apache.geode.redis.ConcurrentLoopingThreads;
+import org.apache.geode.test.awaitility.GeodeAwaitility;
+import org.apache.geode.test.dunit.rules.MemberVM;
+import org.apache.geode.test.dunit.rules.RedisClusterStartupRule;
+
+public class HExistsDUnitTest {
+
+  @ClassRule
+  public static RedisClusterStartupRule clusterStartUp = new RedisClusterStartupRule(4);
+
+  private static final String LOCAL_HOST = "127.0.0.1";
+  private static final int HASH_SIZE = 1000;
+  private static final int JEDIS_TIMEOUT =
+      Math.toIntExact(GeodeAwaitility.getTimeout().toMillis());
+  private static Jedis jedis1;
+  private static Jedis jedis2;
+  private static Jedis jedis3;
+
+  private static Properties locatorProperties;
+
+  private static MemberVM locator;
+  private static MemberVM server1;
+  private static MemberVM server2;
+
+  private static int redisServerPort1;
+  private static int redisServerPort2;
+
+  @BeforeClass
+  public static void classSetup() {
+    locatorProperties = new Properties();
+    locatorProperties.setProperty(MAX_WAIT_TIME_RECONNECT, "15000");
+
+    locator = clusterStartUp.startLocatorVM(0, locatorProperties);
+    server1 = clusterStartUp.startRedisVM(1, locator.getPort());
+    server2 = clusterStartUp.startRedisVM(2, locator.getPort());
+
+    redisServerPort1 = clusterStartUp.getRedisPort(1);
+    redisServerPort2 = clusterStartUp.getRedisPort(2);
+
+    jedis1 = new Jedis(LOCAL_HOST, redisServerPort1, JEDIS_TIMEOUT);
+    jedis2 = new Jedis(LOCAL_HOST, redisServerPort2, JEDIS_TIMEOUT);
+    jedis3 = new Jedis(LOCAL_HOST, redisServerPort1, JEDIS_TIMEOUT);
+  }
+
+  @Before
+  public void testSetup() {
+    jedis1.flushAll();
+  }
+
+  @AfterClass
+  public static void tearDown() {
+    jedis1.disconnect();
+    jedis2.disconnect();
+
+    server1.stop();
+    server2.stop();
+  }
+
+  @Test
+  public void testConcurrentHExists_whileUpdatingValues() {
+    String key = "key";
+
+    Map<String, String> testMap = makeHashMap(HASH_SIZE, "field-", "value-");
+
+    jedis1.hset(key, testMap);
+
+    new ConcurrentLoopingThreads(HASH_SIZE,
+        (i) -> jedis1.hset(key, "field-" + i, "changedValue-" + i),
+        (i) -> assertThat(jedis2.hexists(key, "field-" + i)).isTrue(),
+        (i) -> assertThat(jedis3.hexists(key, "field-" + i)).isTrue()).run();
+
+    Map<String, String> expectedResult = makeHashMap(HASH_SIZE, "field-", "changedValue-");
+    assertThat(jedis1.hgetAll(key)).containsExactlyInAnyOrderEntriesOf(expectedResult);
+  }
+
+  @Test
+  public void testConcurrentHExists_whileAddingValues() {
+    String key = "key";
+
+    Map<String, String> expectedValues = new HashMap<>();
+
+    new ConcurrentLoopingThreads(HASH_SIZE,
+        (i) -> {
+          jedis1.hset(key, "field-" + i, "value-" + i);
+          expectedValues.put("field-" + i, "value-" + i);
+        },
+        (i) -> GeodeAwaitility.await().atMost(Duration.ofSeconds(1))
```

Comment:
This would be more efficient using a queue to pass the field name from one thread to another (as we had done in some other tests) - the test will run way quicker. It also removes the notion that there is some kind of eventual consistency at play here.

Reply:
the downside to doing it with the queue is that it makes it essentially synchronous

Reply:
With a queue, there's never a time when we're calling HEXISTS on something that doesn't exist yet, right? Awaitility is definitely slower, but does seem to allow more "asynchronicity".

I suppose an alternative *might* be randomly HEXISTSing fields in the appropriate range. But at that point, we couldn't know if the return value was really correct, right?